### ISSUE TITLE
Abnormal behavior in mAP calculation 

### ISSUE THREAD

**Issue by lodm94** [PyTorchLightning/metrics#881]
## 🐛 Bug

MeanAveragePrecision from torchmetrics.detection.map exhibits abnormal results when there are False Positives or empty GT/predictions.
I'm running the latest version of torchmetrics.

### Code sample to Reproduce
```py
import torch
from torchmetrics.detection.map import MeanAveragePrecision
preds = [
  dict(
    boxes=torch.Tensor([]),
    scores=torch.Tensor([]),
    labels=torch.IntTensor([]),
  ),
  dict(
    boxes=torch.Tensor([[258.0, 41.0, 606.0, 285.0],[10000.0, 41000.0, 60006.0, 28005.0]]),
    scores=torch.Tensor([0.356,0.32]),
    labels=torch.IntTensor([0,0]),
  )

]


target = [
  dict(
    boxes=torch.Tensor([[258.0, 41.0, 606.0, 285.0],[10000.0, 41000.0, 60006.0, 28005.0]]),
    labels=torch.IntTensor([0,0]),
  ),
  dict(
    boxes=torch.Tensor([]),
    labels=torch.IntTensor([]),
  )
   
]
metric = MeanAveragePrecision()
metric.update(preds, target)
from pprint import pprint
pprint(metric.compute())
```

This code returns a map of -1!
How is this possible?

I take advantage of this thread to ask another thing: how can i calculate just precision and recall for object detection problem? I saw that the function only holds for classification problem since they require as inputs just the predicted/target classes.

**Comment by Borda** [comment]
@twsl @tkupek ^^ pls :rabbit:

**Comment by tkupek** [comment]
Hi @lodm94 thanks for your contribution.

Regarding the question on classification: Object detection models mostly combine detection (localization) and classification problems. For example the famous [COCO detection challenge](https://cocodataset.org/#detection-2020) looks for bounding boxes and the respective class of the objects.
If you don't classify your objects that is equivalent to only having a single class. You can simply pass class `0` for all targets and predictions to make it work.

Regarding your results (appreciate the short example), it's unclear to me how the mAP value should look like on an empty target. If there is no target, how can we calculate how close we are to the non-existing target?

The concern with the empty prediction is valid and already discussed in this issue: https://github.com/PyTorchLightning/metrics/issues/794
Maybe @mtailanian can give a short update?

All the best!

**Comment by MatiTaila** [comment]
Hi @tkupek and @lodm94,

@lodm94 You have a point. As mentioned in issue #794, the problem is that the metric calculation is not taking into account the images that do not have any detection or any ground truth. Therefore, in your code example, you are computing the metric over 0 images, and it returns the default initialization value, -1. It means it couldn't compute the metric.

Note that usually object detectors output a fixed amount of boxes (for example 100 boxes per image), probably most of them with very low confidence. So in this case (which I believe is the most common), there is no problem with the metric.

I've been working on a fix to penalize the metric in those cases. I will soon send a PR, and get it reviewed. 

@lodm94 Regarding your other question: you can recover the precision and recall values from inside the metric. Specifically, inside the `_compute` method, in [this line](https://github.com/PyTorchLightning/metrics/blob/7ce70325951215559926593b4281962e2f5bdcab/torchmetrics/detection/map.py#L710), which is the output of the `_calculate` method

Example code for one specific IoU, one specific box size, class, and max number of detections allowed:
```
target_iou = 0.5  # IoU for which we want to extract the metrics
idx_iou = np.where(self.iou_thresholds.numpy() == target_iou)
idx_cls = 0  # first class
idx_bbox_area = 0  # complete area range, 1 is small, 2 is medium, 3 is large
max_det = np.max(self.max_detection_thresholds.numpy())
idx_max_det_thrs = np.where(self.max_detection_thresholds == max_det)[0]

p = precision[idx_iou, :, idx_cls, idx_bbox_area, idx_max_det_thrs].ravel()
r = self.rec_thresholds.numpy().ravel()
```
With `p` and ` r` you can draw the precision-recall curve, and get your preferred operation point

Hope this is useful...

Thanks

**Comment by lodm94** [comment]
Hi @mtailanian! Thanks for your kind reply! I would like to show you another strange behavior:
### Input 1
```
preds = [
  dict(
    boxes=torch.Tensor([[1,2,3,4],[5,6,7,8]]),
    scores=torch.Tensor([0.8,0.8]),
    labels=torch.IntTensor([0,0]),
  ),
  dict(
    boxes=torch.Tensor([[1,2,3,4],[5,6,7,8]]),
    scores=torch.Tensor([0.7,0.7]),
    labels=torch.IntTensor([0,0]),
  )
]
target = [
  dict(
    boxes=torch.Tensor([[1,2,3,4],[5,6,7,8]]),
    labels=torch.IntTensor([0,0]),
  ),
  dict(
    boxes=torch.Tensor([]),
    labels=torch.IntTensor([]),
  )
]
metric = MeanAveragePrecision()
metric.update(preds, target)
```
### Output 1
```
{'map': tensor(1.),
 'map_50': tensor(1.),
 'map_75': tensor(1.),
 'map_large': tensor(-1.),
 'map_medium': tensor(-1.),
 'map_per_class': tensor(-1.),
 'map_small': tensor(1.),
 'mar_1': tensor(0.5000),
 'mar_10': tensor(1.),
 'mar_100': tensor(1.),
 'mar_100_per_class': tensor(-1.),
 'mar_large': tensor(-1.),
 'mar_medium': tensor(-1.),
 'mar_small': tensor(1.)}
```

### Input 2
```
preds = [
  dict(
    boxes=torch.Tensor([[1,2,3,4],[5,6,7,8]]),
    scores=torch.Tensor([0.8,0.8]),
    labels=torch.IntTensor([0,0]),
  ),
]
target = [
  dict(
    boxes=torch.Tensor([[1,2,3,4],[5,6,7,8]]),
    labels=torch.IntTensor([0,0]),
  ),
]
metric = MeanAveragePrecision()
metric.update(preds, target)
```

### Output 2:
```
{'map': tensor(1.),
 'map_50': tensor(1.),
 'map_75': tensor(1.),
 'map_large': tensor(-1.),
 'map_medium': tensor(-1.),
 'map_per_class': tensor(-1.),
 'map_small': tensor(1.),
 'mar_1': tensor(0.5000),
 'mar_10': tensor(1.),
 'mar_100': tensor(1.),
 'mar_100_per_class': tensor(-1.),
 'mar_large': tensor(-1.),
 'mar_medium': tensor(-1.),
 'mar_small': tensor(1.)}
```
Isn't it strange that these different inputs produce the same output? The second result is coherent with expectations, but in the first case it seems that metric computation does not take into account two False Positives on the second image (second dict).

Regarding the precision and recall computation, thanks! I will take a look at your solution!

**Comment by tkupek** [comment]
Hi @lodm94,

the behavior you posted is correct. The second prediction is not a "false positive". It's filtered out by NMS because it's the exact same box as the first prediction. This is a standard behaviour in object detection.

Source: https://towardsdatascience.com/non-maximum-suppression-nms-93ce178e177c

This is done because, as mentioned by @mtailanian, object detection models output a lot of candidates (usually around 100), they are merged by NMS and filtered with the detection score. The result after these operations is used to calculate the mAP value.

**Comment by lodm94** [comment]
Yes @mtailanian  it is absolutely clear. 
If you want, you can close the issue, your call!

Best

**Comment by mtailanian** [comment]
Hi @tkupek, I just saw your edited message. Great, thanks!
As @lodm94 suggested, please feel free to close the issue (I cannot do it as I'm not a contributor (I think...)). 
Best!

**Comment by SkafteNicki** [comment]
Closing the issue based on the discussion.
